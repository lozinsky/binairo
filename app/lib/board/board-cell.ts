export type BoardCellKind = 'fixed' | 'regular';

export type BoardCellKindRaw = 0 | 1;

export type BoardCellState = 'B' | 'E' | 'R';

export type BoardCellStateRaw = 0 | 1 | 2;

export function isBoardCellKindRaw(value: unknown): value is BoardCellKindRaw {
  return typeof value === 'number' && value in BOARD_CELL_KIND_BY_RAW;
}

export function isBoardCellStateRaw(value: unknown): value is BoardCellStateRaw {
  return typeof value === 'number' && value in BOARD_CELL_STATE_BY_RAW;
}

const NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellState>> = {
  B: 'E',
  E: 'R',
  R: 'B',
};

const BALANCE_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, number>> = {
  B: -1,
  E: 0,
  R: 1,
};

const RAW_BY_BOARD_CELL_KIND: Readonly<Record<BoardCellKind, BoardCellKindRaw>> = {
  fixed: 0,
  regular: 1,
};

const BOARD_CELL_KIND_BY_RAW: Readonly<Record<BoardCellKindRaw, BoardCellKind>> = {
  0: 'fixed',
  1: 'regular',
};

const RAW_BY_BOARD_CELL_STATE: Readonly<Record<BoardCellState, BoardCellStateRaw>> = {
  B: 0,
  E: 2,
  R: 1,
};

const BOARD_CELL_STATE_BY_RAW: Readonly<Record<BoardCellStateRaw, BoardCellState>> = {
  0: 'B',
  1: 'R',
  2: 'E',
};

export type BoardCellValue = readonly [kind: BoardCellKindRaw, state: BoardCellStateRaw];

export class BoardCell {
  get balance() {
    return BALANCE_BY_BOARD_CELL_STATE[this.#state];
  }

  get isEmpty() {
    return this.#state === 'E';
  }

  get isFilled() {
    return !this.isEmpty;
  }

  get isFixed() {
    return this.#kind === 'fixed';
  }

  get state() {
    return this.#state;
  }

  readonly #kind: BoardCellKind;
  readonly #state: BoardCellState;

  constructor(kind: BoardCellKind, state: BoardCellState) {
    this.#kind = kind;
    this.#state = state;
  }

  static create(state: BoardCellState) {
    return new this('regular', state);
  }

  static from(value: BoardCellValue) {
    return new this(BOARD_CELL_KIND_BY_RAW[value[0]], BOARD_CELL_STATE_BY_RAW[value[1]]);
  }

  equals(other: BoardCell) {
    return this.#state === other.#state;
  }

  next() {
    return this.isFixed ? this : new BoardCell(this.#kind, NEXT_BOARD_CELL_STATE_BY_PREV_BOARD_CELL_STATE[this.#state]);
  }

  toFixed() {
    return new BoardCell('fixed', this.#state);
  }

  valueOf(): BoardCellValue {
    return [RAW_BY_BOARD_CELL_KIND[this.#kind], RAW_BY_BOARD_CELL_STATE[this.#state]];
  }
}

export function isBoardCellValue(value: unknown): value is BoardCellValue {
  if (Array.isArray(value) && value.length === 2) {
    return isBoardCellKindRaw(value[0]) && isBoardCellStateRaw(value[1]);
  }

  return false;
}
