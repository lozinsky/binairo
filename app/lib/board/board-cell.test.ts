import { expect, test } from 'vitest';

import { BoardCell, BoardCellKind, type BoardCellPair, BoardCellState } from './board-cell';

test.each([BoardCellState.B, BoardCellState.E, BoardCellState.R])('returns board cell from state', (state) => {
  expect(BoardCell.create(state)).toMatchSnapshot();
});

test.each([
  [[BoardCellKind.Regular, BoardCellState.B] as BoardCellPair],
  [[BoardCellKind.Regular, BoardCellState.E] as BoardCellPair],
  [[BoardCellKind.Regular, BoardCellState.R] as BoardCellPair],
])('returns board cell from value of', (value) => {
  expect(BoardCell.from(value)).toMatchSnapshot();
});

test.each([
  [BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.B)],
  [BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.E)],
  [BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.R)],
])('returns "true" if one board cell is equal to other', (cell, other) => {
  expect(cell.equals(other)).toMatchSnapshot();
});

test.each([
  [BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.E)],
  [BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.R)],
  [BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R)],
  [BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.B)],
  [BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.E)],
  [BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.B)],
])('returns "false" if one board cell is not equal to other', (cell, other) => {
  expect(cell.equals(other)).toMatchSnapshot();
});

test.each([
  BoardCell.create(BoardCellState.B),
  BoardCell.create(BoardCellState.E),
  BoardCell.create(BoardCellState.R),
  new BoardCell(BoardCellKind.Fixed, BoardCellState.E),
  new BoardCell(BoardCellKind.Fixed, BoardCellState.R),
  new BoardCell(BoardCellKind.Fixed, BoardCellState.B),
])('returns next board cell', (cell) => {
  expect(cell.next()).toMatchSnapshot();
});

test.each([BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R)])(
  'returns fixed board cell',
  (cell) => {
    expect(cell.toFixed()).toMatchSnapshot();
  },
);

test.each([BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R)])(
  'returns board cell value of',
  (cell) => {
    expect(cell.valueOf()).toMatchSnapshot();
  },
);

test.each([BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R)])(
  'returns board cell balance',
  (cell) => {
    expect(cell.balance).toMatchSnapshot();
  },
);

test('returns "true" if board cell is empty', () => {
  expect(BoardCell.create(BoardCellState.E).isEmpty).toMatchSnapshot();
});

test.each([BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.B)])(
  'returns "false" if board cell is not empty',
  (cell) => {
    expect(cell.isEmpty).toMatchSnapshot();
  },
);

test.each([BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.B)])(
  'returns "true" if board cell is filled',
  (cell) => {
    expect(cell.isFilled).toMatchSnapshot();
  },
);

test('returns "false" if board cell is not filled', () => {
  expect(BoardCell.create(BoardCellState.E).isFilled).toMatchSnapshot();
});

test.each([
  new BoardCell(BoardCellKind.Fixed, BoardCellState.E),
  new BoardCell(BoardCellKind.Fixed, BoardCellState.R),
  new BoardCell(BoardCellKind.Fixed, BoardCellState.B),
])('returns "true" if board cell is fixed', (cell) => {
  expect(cell.isFixed).toMatchSnapshot();
});

test.each([BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R), BoardCell.create(BoardCellState.B)])(
  'returns "false" if board cell is not fixed',
  (cell) => {
    expect(cell.isFixed).toMatchSnapshot();
  },
);

test.each([BoardCell.create(BoardCellState.B), BoardCell.create(BoardCellState.E), BoardCell.create(BoardCellState.R)])(
  'returns board cell state',
  (cell) => {
    expect(cell.state).toMatchSnapshot();
  },
);
