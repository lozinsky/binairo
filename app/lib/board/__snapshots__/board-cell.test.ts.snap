// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`returns "false" if board cell is not empty 1`] = `false`;

exports[`returns "false" if board cell is not empty 2`] = `false`;

exports[`returns "false" if board cell is not filled 1`] = `false`;

exports[`returns "false" if board cell is not fixed 1`] = `false`;

exports[`returns "false" if board cell is not fixed 2`] = `false`;

exports[`returns "false" if board cell is not fixed 3`] = `false`;

exports[`returns "false" if one board cell is not equal to other 1`] = `false`;

exports[`returns "false" if one board cell is not equal to other 2`] = `false`;

exports[`returns "false" if one board cell is not equal to other 3`] = `false`;

exports[`returns "false" if one board cell is not equal to other 4`] = `false`;

exports[`returns "false" if one board cell is not equal to other 5`] = `false`;

exports[`returns "false" if one board cell is not equal to other 6`] = `false`;

exports[`returns "true" if board cell is empty 1`] = `true`;

exports[`returns "true" if board cell is filled 1`] = `true`;

exports[`returns "true" if board cell is filled 2`] = `true`;

exports[`returns "true" if board cell is fixed 1`] = `true`;

exports[`returns "true" if board cell is fixed 2`] = `true`;

exports[`returns "true" if board cell is fixed 3`] = `true`;

exports[`returns "true" if one board cell is equal to other 1`] = `true`;

exports[`returns "true" if one board cell is equal to other 2`] = `true`;

exports[`returns "true" if one board cell is equal to other 3`] = `true`;

exports[`returns board cell balance 1`] = `-1`;

exports[`returns board cell balance 2`] = `0`;

exports[`returns board cell balance 3`] = `1`;

exports[`returns board cell from state 1`] = `[ðŸŸ¦]`;

exports[`returns board cell from state 2`] = `[â¬œ]`;

exports[`returns board cell from state 3`] = `[ðŸŸ¥]`;

exports[`returns board cell from value of 1`] = `[ðŸŸ¦]`;

exports[`returns board cell from value of 2`] = `[â¬œ]`;

exports[`returns board cell from value of 3`] = `[ðŸŸ¥]`;

exports[`returns board cell state 1`] = `0`;

exports[`returns board cell state 2`] = `2`;

exports[`returns board cell state 3`] = `1`;

exports[`returns board cell value of 1`] = `
[
  1,
  0,
]
`;

exports[`returns board cell value of 2`] = `
[
  1,
  2,
]
`;

exports[`returns board cell value of 3`] = `
[
  1,
  1,
]
`;

exports[`returns fixed board cell 1`] = `{ðŸŸ¦}`;

exports[`returns fixed board cell 2`] = `{â¬œ}`;

exports[`returns fixed board cell 3`] = `{ðŸŸ¥}`;

exports[`returns next board cell 1`] = `[â¬œ]`;

exports[`returns next board cell 2`] = `[ðŸŸ¥]`;

exports[`returns next board cell 3`] = `[ðŸŸ¦]`;

exports[`returns next board cell 4`] = `{â¬œ}`;

exports[`returns next board cell 5`] = `{ðŸŸ¥}`;

exports[`returns next board cell 6`] = `{ðŸŸ¦}`;
